include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -874,6 +907,321 @@ struct __simple_write_to_id
+    {
+        if constexpr (_CopyDiffSetA)
+        {
+            __temp_out.set(__count, __ele_rng1);
+            ++__count;
+        }
+        ++__idx1;
+        ++__idx;
+    }
+    else if (__comp(__ele_rng2, __ele_rng1))
+    {
+        if constexpr (_CopyDiffSetB)
+        {
+            __temp_out.set(__count, __ele_rng2);
+            ++__count;
+        }
+        ++__idx2;
+        ++__idx;
+    }
+    else // if neither element is less than the other, they are equal
+    {
+        if constexpr (_CopyMatch)
+        {
+            __temp_out.set(__count, __ele_rng1);
+            ++__count;
+        }
+        ++__idx1;
+        ++__idx2;
+        __idx += 2;
+    }
+}
+
+template <bool _CopyMatch, bool _CopyDiffSetA, bool _CopyDiffSetB>
+struct __set_generic_operation
+{
+    template <typename _InRng1, typename _InRng2, typename _SizeType, typename _TempOutput, typename _Compare>
+    std::uint16_t
+    operator()(const _InRng1& __in_rng1, const _InRng2& __in_rng2, std::size_t __idx1, std::size_t __idx2,
+               _SizeType __num_eles_min, _TempOutput& __temp_out, _Compare __comp) const
+    {
+
+        std::uint16_t __count = 0;
+        _SizeType __idx = 0;
+        bool __can_reach_rng1_end = __idx1 + __num_eles_min >= __in_rng1.size();
+        bool __can_reach_rng2_end = __idx2 + __num_eles_min >= __in_rng2.size();
+
+        if (!__can_reach_rng1_end && !__can_reach_rng2_end)
+        {
+            while (__idx < __num_eles_min)
+            {
+                // no bounds checking
+                __set_generic_operation_iteration<_CopyMatch, _CopyDiffSetA, _CopyDiffSetB, false, false>(
+                    __in_rng1, __in_rng2, __idx1, __idx2, __num_eles_min, __temp_out, __idx, __count, __comp);
+            }
+        }
+        else if (__can_reach_rng1_end && __can_reach_rng2_end)
+        {
+            while (__idx < __num_eles_min)
+            {
+                //bounds check all
+                __set_generic_operation_iteration<_CopyMatch, _CopyDiffSetA, _CopyDiffSetB, true, true>(
+                    __in_rng1, __in_rng2, __idx1, __idx2, __num_eles_min, __temp_out, __idx, __count, __comp);
+            }
+        }
+        // bounds check only the range that needs it
+        else if (__can_reach_rng1_end)
+        {
+            while (__idx < __num_eles_min)
+            {
+                __set_generic_operation_iteration<_CopyMatch, _CopyDiffSetA, _CopyDiffSetB, true, false>(
+                    __in_rng1, __in_rng2, __idx1, __idx2, __num_eles_min, __temp_out, __idx, __count, __comp);
+            }
+        }
+        else //(__can_reach_rng2_end)
+        {
+            while (__idx < __num_eles_min)
+            {
+                __set_generic_operation_iteration<_CopyMatch, _CopyDiffSetA, _CopyDiffSetB, false, true>(
+                    __in_rng1, __in_rng2, __idx1, __idx2, __num_eles_min, __temp_out, __idx, __count, __comp);
+            }
+        }
```

Comment:
Note to self: We may want to consider just bounds check or no bounds check, and not have all these permutations, as they are also runtime branches.  
Compare performance.

Reply:
Got an extra ~5% performance overall from this change, applied.